### ISSUE TITLE
JsonSerializer.Deserialize<T> cannot serialize into custom IEnumerable

### ISSUE THREAD

**Issue by HughPH** [dotnet/corefx#41105]
My class is designed to get the items from an Amazon DynamoDB query and convert them to T through the magic of the Enumerator:

```csharp
    public class DynamoQueryResult<T> : IEnumerable<T> where T : class, new()
    {
        public int Count;
        public Dictionary<string, DynamoValue>[] Items;
        public IEnumerator<T> GetEnumerator()
        {
            return new DynamoQueryResultEnumerator<T>(this);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return new DynamoQueryResultEnumerator<T>(this);
        }
    }
```

Json Deserializers detect IEnumerable and expect the JSON at that level to therefore be a set (where this class represents a complex object containing both set metadata and set members). I can override that behaviour with Newtonsoft by adding the [JsonObject] attribute to the object and it does the job perfectly - and Newtonsoft's exception message gives a load of help which leads to this. I can't find a way to do the same for System.Text.Json. Also, really descriptive exception messages that guide the developer are one of my favourite things to create and see.

1. I know there is a big library of junk for dealing with Amazon stuff. That is not the answer I'm looking for - it's likely to just add delay and potentially more questions. This is a working _example_ of where this falls over.
2. In the immediate term, I'm likely to just create a customised collection to replace the array of dictionaries, but it would be cool if I could just convince the deserializer to fill the object.

**Comment by layomia** [comment]
>ReadOnly IEnumerables could be a way to 'hint' that they're not meant to be deserialized into

I don't think this is enough information for the serializer to change the normal behavior.

It seems that the way to move forward here is if there were an explicit way to tell the serializer to treat an `IEnumerable` like an object with properties, similar to Netonsoft's [JsonObject] attribute. I'll rename this issue to reflect this.
